Change PathLikeWithPosition<P> into a non-generic type and replace ad-hoc Windows path parsing

sndels opened the pull request:
This simplifies `PathWithPosition` by making the common use case concrete and removing the manual, incomplete Windows path parsing. Windows paths also don't get '/'s replaced by '\\'s anymore to limit the responsibility of the code to just parsing out the suffix and creating `PathBuf` from the rest. `Path::file_name()` is now used to extract the filename and potential suffix instead of manual parsing from the full input. This way e.g. Windows paths that begin with a drive letter are handled correctly without platform-specific hacks.

Release Notes:

- N/A

SomeoneToIgnore reviewed (changes requested):
I somehow missed the whole `fn parse_absolute_path<E>(` appearance which is sad, and I do not understand the purpose of both that and this change.

The whole `PathLikeWithPosition` struct exists for the sole purpose of parsing away `PATH` from `:something:else` suffixes (https://github.com/zed-industries/zed/issues/15412 would add another,  `(row,column)` , one).

Would it be not a suffix but a prefix, the key point is that the `PATH` component stays the `PathBuf` anyway. We can remove the generic and enforce it on a type level, if needed.

And Rust's stdlib is already capable of parsing `PATH` the correct way, we do not seem to need all this code at all?

So, instead of reproducing the `PathBuf::from()` for `String`, can we reuse that instead for Windows?

sndels commented:
It does sound like a good idea to use the existing `PathBuf` parse for the path part and make the type concrete if that's the only use case. Do you mean that `parse_str()` would try `PathBuf::from()` first and then try to parse the potential remainder for row-col suffix with whatever formats are supported? Splitting after the already parsed path would get around ':' being both a drive separator and split in the :row:colum: suffix more nicely than the manual hacks.

sndels commented:
Looks like `PathBuf::from` takes the input as is so here it has whatever suffix is given. We could try to parse suffix from `path_buf.as_path().file_name()` and replace the filename with whatever comes before the suffix if our parsing succeeds. Fallback would be the full input as `PathBuf` and it may or may not be a valid path.

sndels commented:
`file_finder` seems to be using this with `FileSearchQuery` as the concrete type so swapping the generic for `PathBuf` isn't as clear-cut. We can get the filename/final part parsing with `Path::from(&str)` while keeping the current interface. I'll try that approach to get rid of the target windows parts without changing how this behaves in general.

SomeoneToIgnore commented:
I just wonder could we do the 

```
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, Hash)]
pub struct PathLikeWithPosition {
    pub path: PathBuf,
    pub row: Option<u32>,
    // Absent if row is absent.
    pub column: Option<u32>,
}
```
given that we already do
https://github.com/zed-industries/zed/blob/4020f64786231f19a80a87cce8676d85360051a6/crates/util/src/paths.rs#L144-L145
and `path_like_str.trim()` is the thing we could turn into path various ways, e.g. `PathBuf::from(path_like_str.to_owned())`.

After this change, the name could be better too.

As the rest of the code that uses the custom thing is in `file_finder.rs` only, `FileSearchQuery` could wrap the `PathLikeWithPosition` instead (or, maybe there's even a better way to represent types in this case).

SomeoneToIgnore commented:
And if that works (I see no reasons why I cannot), all the Windows-related path parsing code could be thrown away.

sndels commented:
That sounds reasonable, including the inversion of `FileSearchQuery`. I'll try this route then.

sndels committed:
file_finder: Invert PathLikeWithPosition<FileSearchQuery>

This was the only place not using PathLikeWithPosition<PathBuf>. We're
making PathWithPosition into a non-generic type that holds a PathBuf to
simplify its internals so let's have FileSearchQuery hold path_position
instead.

sndels committed:
Change PathLikeWithPosition<P> into PathWithPosition

All users had PathBuf as the generic type and having it be concrete
makes parsing internals simpler. Also remove the first argument of the
parse_str closure because it was only used by tests and even that seemed
redundant.

sndels committed:
PathWithPosition: Remove ad-hoc Windows path parsing

We can extract the filename and suffix using `Path` and share the suffix
parsing code without needing to resort to platform specific hacks. This
should make the parsing agnostic to any prefixes in the paths.

One observable difference to the previously working behavior is that '/'s
are no longer replaced by '\'s on Windows. The role of PathWithPosition
is only to parse out the row-col suffix if it exists and create a PathBuf
from the rest of the input string.

sndels commented:
Yeah, having `FileSearchQuery` to contain `path_position` instead and then making the type a concrete `PathWithPosition` ended up being pretty clean. Using `Path` to get to the combined filename and suffix before suffix parsing seemed to work as expected too, letting me remove all the windows target code while getting more path types working "for free".

I'll comb through the changes once more tomorrow before force pushing them and renaming this issue to reflect what's being done. Or should I make a whole new PR instead?

SomeoneToIgnore commented:
I think it's fine to continue here, but up to you.
Ping me when ready then.

sndels commented:
There. The one reservation I have is that the new semantics of the type might suggest more strongly that it is a valid path, but it's probably ok since `PathBuf` and `Path` aren't guaranteed to be valid paths either.

sndels commented:
No diff in that last push because I fixed a fixup mishap between the first two commits.

SomeoneToIgnore commented on crates/util/src/paths.rs:
```diff
@@ -96,59 +96,61 @@ pub const FILE_ROW_COLUMN_DELIMITER: char = ':';
     pub fn parse_str<E>(
         s: &str,
-        parse_path_like_str: impl Fn(&str, &str) -> Result<P, E>,
+        parse_path_str: impl Fn(&str) -> Result<PathBuf, E>,
```
Let's remove this one too, as it's doing the very same fallback across all the codebase: parse the original string as a path.

SomeoneToIgnore reviewed (commented):
Ganska bra!

One last thing to do and we're ready to merge.

sndels committed:
Remove PathWithPosition::parse_str() closure argument

All callers converted into PathBuf without special handling so let's
always do that internally. This also removes the need for the Result
wrapper on return.

SomeoneToIgnore reviewed (approved):
Thank you for cleaning this up, looks much better than before!
